Sources/SwiftSyntax/Utils.swift

```diff
@@ -10,62 +10,86 @@
+public typealias ByteSourceRange = Range<AbsolutePosition>
 
-  public init(offset: Int, length: Int) {
-    self.offset = offset
-    self.length = length
+public extension Range<AbsolutePosition> {
+  @available(*, deprecated, message: "Use lowerBound.utf8Offset instead")
+  var offset: Int { lowerBound.utf8Offset }
+
+  @available(*, deprecated, message: "Construct a Range<AbsolutePosition> instead")
+  init(offset: Int, length: Int) {
+    self = AbsolutePosition(utf8Offset: offset)..<AbsolutePosition(utf8Offset: offset + length)
   }
 
-  public var endOffset: Int {
-    return offset + length
+  @available(*, deprecated, message: "Use upperBound.utf8Offset instead")
+  var endOffset: Int {
+    return upperBound.utf8Offset
   }
 
-  public var isEmpty: Bool {
-    return length == 0
+  /// Returns the range for the overlapping region between two ranges.
+  ///
+  /// If the intersection is empty, this returns a range starting at offset 0 and having length 0.
+  func intersected(_ other: Range<AbsolutePosition>) -> Range<AbsolutePosition> {
+    return self.intersecting(other) ?? AbsolutePosition(utf8Offset: 0)..<AbsolutePosition(utf8Offset: 0)
+  }
+}
+
+extension Range<AbsolutePosition> {
+  /// The number of bytes between the range's lower bound and its upper bound
+  public var length: SourceLength { return SourceLength(utf8Length: upperBound.utf8Offset - lowerBound.utf8Offset) }
+
+  public init(position: AbsolutePosition, length: SourceLength) {
+    self = position..<(position + length)
   }
 
-  public func intersectsOrTouches(_ other: ByteSourceRange) -> Bool {
-    return self.endOffset >= other.offset && self.offset <= other.endOffset
+  /// Returns `true` if the intersection between this range and `other` is non-empty or if the two ranges are directly
+  /// adjacent to each other.
+  public func intersectsOrTouches(_ other: Range<AbsolutePosition>) -> Bool {
+    return self.upperBound >= other.lowerBound && self.lowerBound <= other.upperBound
   }
 
-  public func intersects(_ other: ByteSourceRange) -> Bool {
-    return self.endOffset > other.offset && self.offset < other.endOffset
+  /// Returns `true` if the intersection between this range and `other` is non-empty.
+  public func intersects(_ other: Range<AbsolutePosition>) -> Bool {
+    return self.upperBound > other.lowerBound && self.lowerBound < other.upperBound
   }
 
-  /// Returns the byte range for the overlapping region between two ranges.
-  public func intersected(_ other: ByteSourceRange) -> ByteSourceRange {
-    let start = max(self.offset, other.offset)
-    let end = min(self.endOffset, other.endOffset)
-    if start > end {
-      return ByteSourceRange(offset: 0, length: 0)
+  /// Returns the range for the overlapping region between two ranges.
+  ///
+  /// If the intersection is empty, this returns `nil`.
+  public func intersecting(_ other: Range<AbsolutePosition>) -> Range<AbsolutePosition>? {
```

Comment:
I feel like we should use `Range` methods like  `clamped(_:)` / `overlaps(_:)` . Is there a reason we can't?

Reply:
Very good suggestion. Added a commit that deprecates `intersection`-style methods.